include/experimental/__p2642_bits/layout_padded.hpp

```diff
@@ -0,0 +1,808 @@
+  extents_type __extents = {};
+
+  constexpr index_type __compute_offset(std::index_sequence<>) const {
+    return 0;
+  }
+
+  template <size_t Rank, class IndexOffset>
+  constexpr index_type __compute_offset(std::index_sequence<Rank>,
+                                        IndexOffset __index_offset) const {
+    return __index_offset;
+  }
+
+  template <size_t... Ranks, class... IndexOffsets>
+  constexpr index_type __compute_offset(std::index_sequence<Ranks...>,
+                                        IndexOffsets... __index_offsets) const {
+    index_type __indices[] = {static_cast<index_type>(__index_offsets)...};
+    // self-recursive fold trick from
+    // https://github.com/llvm/llvm-project/blob/96e1914aa2e6d8966acbfbe2f4d184201f1aa318/libcxx/include/__mdspan/layout_left.h#L144
+    index_type res = 0;
+    ((res = __indices[extents_type::rank() - 1 - Ranks] +
+            ((extents_type::rank() - 1 - Ranks) == __extent_to_pad_idx
+                 ? __padded_stride.value(0)
+                 : __extents.extent(extents_type::rank() - 1 - Ranks)) *
+                res),
+     ...);
+    return res;
+  }
+
+public:
+#if !MDSPAN_HAS_CXX_20
+  MDSPAN_INLINE_FUNCTION_DEFAULTED
+  constexpr mapping()
+      : mapping(extents_type{})
+  {}
+#else
+  MDSPAN_INLINE_FUNCTION_DEFAULTED
+  constexpr mapping()
+    requires(__actual_padding_value != dynamic_extent) = default;
+
+  MDSPAN_INLINE_FUNCTION
+  constexpr mapping()
+    requires(__actual_padding_value == dynamic_extent)
+      : mapping(extents_type{})
+  {}
+#endif
+
+  MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mapping(const mapping&) noexcept = default;
+  MDSPAN_INLINE_FUNCTION_DEFAULTED mapping& operator=(const mapping&) noexcept = default;
+
+  /**
+   * Initializes the mapping with the given extents.
+   *
+   * \param ext the given extents
+   */
+  MDSPAN_INLINE_FUNCTION
+  constexpr mapping(const extents_type& __ext)
+    : __padded_stride(__padded_stride_type::__impl_init_padding(__ext)), __extents(__ext)
+  {}
+
+  /**
+   * Initializes the mapping with the given extents and the specified padding value.
+   *
+   * This overload participates in overload resolution only if `is_convertible_v<Size, index_type>`
+   * is `true` and `is_nothrow_constructible_v<index_type, Size>` is `true`
+   *
+   * \param ext the given extents
+   * \param padding_value the padding value
+   */
+  MDSPAN_TEMPLATE_REQUIRES(
+    class _Size,
+    /* requires */ (
+      std::is_convertible_v<_Size, index_type>
+      && std::is_nothrow_constructible_v<index_type, _Size>
+    )
+  )
+  MDSPAN_INLINE_FUNCTION
+  constexpr mapping(const extents_type &__ext, _Size __dynamic_padding_value)
+      : __padded_stride(__padded_stride_type::__impl_init_padding(__ext, __dynamic_padding_value)), __extents(__ext)
+  {
+    assert((padding_value == dynamic_extent) || (padding_value == static_cast<index_type>(__dynamic_padding_value)));
```

Comment:
The casting to index_type does not feel right.  Why are we not casting to `size_t`?

Reply:
as all other things in mdspan the argument is only guaranteed to be `index_type`. In fact for this comparison we should cast `padding_value` to `index_type`.